pkg/controller/directimagestreammigration/validate.go

```diff
@@ -0,0 +1,306 @@
+	liberr "github.com/konveyor/controller/pkg/error"
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	migref "github.com/konveyor/mig-controller/pkg/reference"
+	kapi "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"reflect"
+)
+
+// Types
+const (
+	InvalidSourceClusterRef               = "InvalidSourceClusterRef"
+	InvalidDestinationClusterRef          = "InvalidDestinationClusterRef"
+	InvalidSourceCluster                  = "InvalidSourceCluster"
+	InvalidDestinationCluster             = "InvalidDestinationCluster"
+	MissingSourceClusterRegistryPath      = "MissingSourceClusterRegistryPath"
+	MissingDestinationClusterRegistryPath = "MissingDestinationClusterRegistryPath"
+	SourceClusterNotReady                 = "SourceClusterNotReady"
+	DestinationClusterNotReady            = "DestinationClusterNotReady"
+	InvalidImageStreamRef                 = "InvalidImageStreamRef"
+	InvalidImageStream                    = "InvalidImageStream"
+	NsNotFoundOnDestinationCluster        = "NamespaceNotFoundOnDestinationCluster"
+
+	Running   = "Running"
+	Failed    = "Failed"
+	Succeeded = "Succeeded"
+)
+
+// Reasons
+const (
+	NotFound    = "NotFound"
+	NotSet      = "NotSet"
+	NotDistinct = "NotDistinct"
+	NotReady    = "NotReady"
+)
+
+// Messages
+const (
+	ReadyMessage                                 = "Image migration is ready"
+	RunningMessage                               = "Step: %d/%d"
+	InvalidSourceClusterReferenceMessage         = "The source cluster reference is invalid"
+	InvalidDestinationClusterReferenceMessage    = "The destination cluster reference is invalid"
+	InvalidSourceClusterMessage                  = "The source cluster is invalid"
+	InvalidDestinationClusterMessage             = "The destination cluster is invalid"
+	SourceClusterNotReadyMessage                 = "The source cluster is not ready"
+	DestinationClusterNotReadyMessage            = "The destination cluster is not ready"
+	MissingSourceClusterRegistryPathMessage      = "The source cluster is missing an exposed registry path"
+	MissingDestinationClusterRegistryPathMessage = "The destination cluster is missing an exposed registry path"
+	InvalidImageStreamReferenceMessage           = "The ImageStream reference is invalid"
+	InvalidImageStreamMessage                    = "The ImageStream is invalid"
+	NsNotFoundOnDestinationClusterMessage        = "Namespace [] not found on the destination cluster."
+	SucceededMessage                             = "The migration has succeeded"
+	FailedMessage                                = "The migration has failed.  See: Errors."
+)
+
+// Categories
+const (
+	Critical = migapi.Critical
+	Advisory = migapi.Advisory
+)
+
+// Statuses
+const (
+	True  = migapi.True
+	False = migapi.False
+)
+
+// Validate the image migration resource
+func (r ReconcileDirectImageStreamMigration) validate(imageStreamMigration *migapi.DirectImageStreamMigration) error {
```

Comment:
A general question about all critical conditions added in this function/sub-functions.

If there is a critical condition added, when will the controller requeue this CR?

It looks like the current code only requeues on unexpected errors, https://github.com/konveyor/mig-controller/pull/710/commits/99fb253d8d8510038f614ae61e3a43b6d5ccf743#diff-775a50c0d704007e87759e873ac5f70e4584cd421209ee5214478461c18b3848R136

If you agree that we need to requeue on critical conditions, there might be two options:
1) Either we make all critical conditions return errors so we requeue
2) we requeue any critical condition present on the CR at the end of the reconcile function.

Reply:
That's a question for @djwhatle or @jortel I think.

Reply:
This logic may differ across each controller. I can't give a generic answer without looking at each controller's reconcile func. The migmigration controller requeues itself when it has a crit condition.